lib/report-translator.js

```diff
@@ -261,11 +263,29 @@ module.exports = function createReportTranslator(metadata) {
 
         assertValidNodeInfo(descriptor);
 
+        if (descriptor.messageId) {
+            if (!metadata.messageIds) {
+                throw new TypeError("context.report() called with a messageId, but no messages were present in the rule metadata.");
+            }
+            const id = descriptor.messageId;
+            const messages = metadata.messageIds;
+
+            if (descriptor.message) {
+                throw new TypeError("context.report() called with a message and a messageId. Please only pass one.");
+            }
+            if (!messages || !Object.prototype.hasOwnProperty.call(messages, id)) {
+                throw new TypeError(`context.report() called with a messageId of '${id}' which is not present in the 'messages' config: ${JSON.stringify(messages, null, 2)}`);
+            }
+            descriptor.message = messages[id];
```

Comment:
Can you avoid mutating the descriptor here? I think it would be better to put the `messageId` logic in `interpolate` so that all of the message-handling logic is in the same place, and the message doesn't need to be reassigned.

(Unlike https://github.com/eslint/eslint/pull/9165#discussion_r135395956, this time I'm not concerned about mutating objects from rules, since I see you cloned the descriptor above. I just think not mutating a function's arguments in general is a good idea for code cleanliness, since it makes code easier to follow.)

Reply:
Hm, this isn’t actually a function argument:

```js
// lib/report-translator.js:262
const descriptor = normalizeMultiArgReportCall.apply(null, arguments);
````

Reply:
Oops, you're right. I still think it would be slightly better to avoid mutation, but I don't feel strongly about it.